keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/AuthorizationSupport.kt

```diff
@@ -18,59 +18,171 @@
+          if (!allowed) {
+            throw AccessDeniedException("User ${auth.principal} does not have access to service account $serviceAccount")
+          }
+        }
+    }
   }
 
-  fun userCanAccessServiceAccount(auth: Authentication, serviceAccount: String, specOrName: Any): Boolean {
-    val hasPermission = permissionEvaluator.hasPermission(auth, serviceAccount, "SERVICE_ACCOUNT", "ignored-svcAcct-auth")
-    log.debug(
-      "[AUTH] {} is trying to access service account {}. They{} have permission. {}",
-      auth.principal,
-      serviceAccount,
-      if (hasPermission) "" else " DO NOT",
-      specOrName
-    )
-    return hasPermission
+  /**
+   * Verifies that the caller has the specified permission to all applicable resources (i.e. resources whose specs
+   * are [Locatable]) identified by the target type and identifier, as follows:
+   *   - If target is RESOURCE, check the resource itself
+   *   - If target is DELIVERY_CONFIG, check all the resources in all the environments of the delivery config
+   *   - If target is APPLICATION, do the same as for DELIVERY_CONFIG
+   *
+   * @throws AccessDeniedException if caller does not have the required permission.
+   */
+  fun hasCloudAccountPermission(action: Action, target: TargetEntity, identifier: String) {
+    if (!enabled()) return
+
+    withAuthentication(target, identifier) { auth ->
+      val locatableResources = when (target) {
+        TargetEntity.RESOURCE -> listOf(repository.getResource(identifier))
+        TargetEntity.APPLICATION -> repository.getDeliveryConfigForApplication(identifier).resources
+        TargetEntity.DELIVERY_CONFIG -> repository.getDeliveryConfig(identifier).resources
+      }.filter { it.spec is Locatable<*> }
+
+      locatableResources.forEach {
+        val account = (it.spec as Locatable<*>).locations.account
+        permissionEvaluator.hasPermission(auth, account, "ACCOUNT", action.name)
+          .also { allowed ->
+            log.debug("[ACCESS {}] User {}: {} access to cloud account {}.", allowed.str, auth.principal, action.name, account)
+            if (!allowed) {
+              throw AccessDeniedException("User ${auth.principal} does not have access to cloud account $account")
+            }
+          }
+      }
+    }
   }
+
+  private fun withAuthentication(target: TargetEntity, identifier: String, block: (Authentication) -> Unit) {
+    try {
+      val auth = SecurityContextHolder.getContext().authentication
+      block(auth)
+    } catch (e: NoSuchEntityException) {
+      // If entity doesn't exist return true so a 404 is returned from the controller.
+      log.debug("${target.name} $identifier not found. Allowing request to return 404.")
+    }
+  }
+
+  private fun passes(authorizationCheck: () -> Unit) =
+    try {
+      authorizationCheck()
+      true
+    } catch (e: AccessDeniedException) {
+      false
+    }
+
+  val Boolean.str: String
```

Comment:
what is this magic?? can you add some comments / name it better for future reference 🙏?

Reply:
No magic, really. I just wanted something that could a `Boolean` that is `true` as `"ALLOWED"`, and `"DENIED"` otherwise to make it easier to log above. Unfortunately, you cannot override the `toString()` method with extensions like this, so I used this `str` property instead.

Reply:
I definitely feel like the name could be more meaningful… `toAuthorization` maybe?